### PR TITLE
Explore transporter

### PR DESCRIPTION
**Pull request by davidpmccormick**
## What is this PR trying to achieve?
Adding the digital story transporter to the explore page.

## What does it look like?
screen shot 2017-05-19 at 11 52 09

### CODE REVIEW

File: server/controllers/index.js [hunk 1]
```diff
@@ -182,12 +182,20 @@ export const explore = async(ctx, next) => {
   };
 
   const aDropInTheOceanPromoList = PromoListFactory.fromSeries(aDropInTheOceanSeries);
+  const digitalStories = commissionedSeries.filter(s => s.commissionedLength);
```
Comment (davidpmccormick) [inline server/controllers/index.js]: I'm not sure if this is the best way to find the most recent digital story – sorting the items of all the stories by date, taking the most recent one, then finding the story that _that_ one lives in. Other/better ways of doing this welcomed.

File: client/scss/components/_slider.scss [hunk 2]
```diff
@@ -23,14 +23,16 @@ $slider-control-margin: 18px;
...
+  &.slider__slider--with-wobbly-line {
+    &:before {
+      height: 60px;
+      display: block;
+      width: 100%;
+      content: '';
+      background: color('white');
+      margin-bottom: -5px;
+      clip-path: polygon(0% 75%, 0% 100%, 100% 100%, 100% 60%, 92% 85%, 70% 67%, 57% 90%, 45% 59%, 20% 72%);
+    }
```
Comment (jamesgorrie) [inline client/scss/components/_slider.scss]: Is this something we could generify as we use it all over the place?

File: client/scss/components/_slider.scss [hunk 3]
```diff
@@ -55,25 +57,37 @@ $slider-control-margin: 18px;
...
   }
 
   @include respond-to('large') {
     width: calc(100vw - #{map-get($container-padding, 'large')});
-    margin-left: 0;
+
+    .slider__slider--in-article & {
+      margin-left: 0;
+    }
```
Comment (jamesgorrie) [inline client/scss/components/_slider.scss]: Having parent specificity always . scares me a little, and isn't very BEM (which should be as flat as possible).
Could we not just add a modifier `slider__slides-container--no-margin` instead?

File: server/views/pages/article.njk [hunk 4]
```diff
@@ -76,12 +76,14 @@
...
-            <div class="body-part body-part--full-width">
-              {% set asyncContentEndpoint = '/series-transporter/' + series.url + '?current=' + article.url %}
-              {% component 'async', { endpoint: asyncContentEndpoint, component: 'series-transporter' } %}
+      <div class="row {{ {s:4, m:4, l:4} | spacingClasses({padding: ['bottom']}) }}">
+        <div class="container digital-story-transporter">
+          <div class="grid grid--no-gutters">
+            <div class="{{ {s: 4, m: 6, shiftM: 1, l: 7, xl: 6, shiftXl: 1} | gridClasses }}">
+              <div class="body-part body-part--full-width">
+                {% set asyncContentEndpoint = '/series-transporter/' + series.url + '?current=' + article.url %}
+                {% component 'async', { endpoint: asyncContentEndpoint, component: 'series-transporter', modifiers: '{"in-article": true, "with-wobbly-line": true}' } %}
```
Comment (jamesgorrie) [inline server/views/pages/article.njk]: Could we not do something in the logic of the component to add the modifiers?
This seems like a very specific piece of functionality wedged onto a generic component.

File: server/views/components/async/index.njk [hunk 5]
```diff
@@ -1 +1 @@
-<div data-component="{{ component }}" class="async-content {{ component }}-placeholder" data-endpoint="{{ endpoint }}"></div>
+<div data-component="{{ component }}" class="async-content {{ component }}-placeholder" data-endpoint="{{ endpoint }}" data-modifiers="{{ modifiers }}"></div>
```
Comment (jamesgorrie) [inline server/views/components/async/index.njk]: This feels like a very dirty place for this.

File: server/controllers/index.js (hunk 1 above)
Comment (jamesgorrie) [inline server/controllers/index.js]: Run me through what you're doing here in person perhaps?
This is a strange part of the controller anyway, so potentially fine.

File: server/views/components/async/index.njk (hunk 5 above)
Comment (davidpmccormick) [inline server/views/components/async/index.njk]: Agreed! Wasn't sure how else to get modifiers into the component.

File: client/scss/components/_slider.scss (hunk 3 above)
Comment (davidpmccormick) [inline client/scss/components/_slider.scss]: Yeah, fair – will update.

File: client/js/components/content-slider.js [hunk 6]
```diff
@@ -75,6 +76,11 @@ const contentSlider = (el, options) => {
     sliderElements.slider.className = classes.slider;
     sliderElements.sliderInner.className = classes.sliderInner;
     sliderElements.slidesContainer.classList.add(classes.slidesContainer);
+    slidesContainerModifiers.split(' ').forEach((modifier) => {
+      if (modifier.trim() !== '') {
```
Comment (jamesgorrie) [inline client/js/components/content-slider.js]: Why do we need this check? We never pass empty strings in?

Comment (jamesgorrie) [inline client/js/components/content-slider.js]: @davidpmccormick just checking on this?

File: client/scss/components/_slider.scss (hunk 2 above)
Comment (jamesgorrie) [inline client/scss/components/_slider.scss]: Could we generify this?
If you think the wobbly line can be pulled out - can we make a note of it somewhere to make sure we do it (on the conversation of reducing complexity).

File: client/scss/components/_slider.scss [hunk 7]
```diff
@@ -55,25 +57,37 @@ $slider-control-margin: 18px;
...
-    margin-left: 50%;
-    left: (($container-width-max - (map-get($container-padding, 'large') * 2)) / 2) * (-5/6);
+    width: calc(100vw - ((100vw - #{$container-width-max}) / 2) - #{map-get($container-padding, 'large')});
+    left: calc((100vw - #{$container-width-max}) / 2);
+
+    &.slider__slides-container--in-article {
+      width: $container-width-max - (map-get($container-padding, 'large') * 2);
+      margin-left: 50%;
+      left: (($container-width-max - (map-get($container-padding, 'large') * 2)) / 2) * (-5/6);
+    }
```
Comment (jamesgorrie) [inline client/scss/components/_slider.scss]: Let's have a chat with @Norvard about all the different ways we seem to do this.
Hopefully we can find one generic class to do this as we seem to do it in loads of places.

File: server/views/components/async/index.njk (hunk 5 above)
Comment (jamesgorrie) [inline server/views/components/async/index.njk]: Could we please leave a TODO here so if we do use it again we question it?

File: server/views/pages/explore.njk [hunk 8]
```diff
@@ -28,6 +28,22 @@
     </div>
   </div>
 
+  <div class="row {{ {s:2, m:2, l:2} | spacingClasses({padding:['top']}) }}">
+    <div class="container container--scroll-xl">
+      <div class="grid grid--no-gutters">
+        <div class="{{ {s: 4, m: 8, l: 12, xl: 12} | gridClasses }}">
+          {% set asyncContentEndpoint = '/series-transporter/' + latestDigitalStory %}
+          {% component 'async', { endpoint: asyncContentEndpoint, component: 'series-transporter' } %}
```
Comment (jamesgorrie) [inline server/views/pages/explore.njk]: So far in mixpanel we've used component name = "Series transporter".
I personally prefer that as it's more readable.

But happy to choose this as a convention, but let's document the convention somewhere.

Comment (davidpmccormick) [inline server/views/pages/explore.njk]: Do you mean titlecase and non-hyphenated? This ('series-transporter') isn't what's being sent to Mixpanel – it is 'Transporter', as determined [in the numbered-list](https://github.com/wellcometrust/wellcomecollection.org/blob/master/server/views/components/numbered-list/index.njk#L15)

Comment (jamesgorrie) [inline server/views/pages/explore.njk]: Ah - [we have a same named thing here](https://github.com/wellcometrust/wellcomecollection.org/pull/911/files#diff-f34dd8a341cd717c49b4bf003a14951bR1) - what are we using that for?

File: client/js/components/content-slider.js (hunk 6 above)
Comment (davidpmccormick) [inline client/js/components/content-slider.js]: This was the product of the `setModifiers` reduce above resulting in a string starting in a space. Have now, more appropriately, trimmed it there instead.

File: client/js/components/content-slider.js [hunk 9]
```diff
@@ -38,7 +38,7 @@ const contentSlider = (el, options) => {
   function setModifiers(cssBlock) {
     return settings.modifiers.reduce((acc, curr) => {
       return `${acc} ${settings.cssPrefix}${cssBlock}--${curr}`;
-    }, '');
+    }, '').trim();
```
Comment (jamesgorrie) [inline client/js/components/content-slider.js]: Still not sure why we need this - why are we sending untrimmed strings?

Comment (davidpmccormick) [inline client/js/components/content-slider.js]: It's the result of a reduction: https://github.com/wellcometrust/wellcomecollection.org/pull/911#discussion_r118221207

File: server/views/pages/explore.njk (hunk 8 above)
Comment (davidpmccormick) [inline server/views/pages/explore.njk]: That `slider-transporter-child` is a modifier for the promo which overrides the [promo's container-query behaviours](https://github.com/wellcometrust/wellcomecollection.org/pull/911/files/73ae975ea95bf4039084cefe8df19a0a7da84ffc#diff-106f0f02dce48cb540307355e721b7e0R386)

Comment (jamesgorrie) [inline server/views/pages/explore.njk]: Any way we could not use the `data-component` attribute and avoid overloading it?

**Pull request merged**